contracts/SignerBondsUniswapV2.sol

```diff
@@ -0,0 +1,260 @@
+    ///        high limit level will accept transactions with high volumes.
+    ///        Those transactions may result in poor execution prices. Very low
+    ///        limit will force low swap volumes. Limit equal to 0 will
+    ///        effectively make swaps impossible.
+    function setMaxAllowedPriceImpact(uint256 _maxAllowedPriceImpact)
+        external
+        onlyOwner
+    {
+        require(
+            _maxAllowedPriceImpact <= BASIS_POINTS_DIVISOR,
+            "Maximum value is 10000 basis points"
+        );
+        maxAllowedPriceImpact = _maxAllowedPriceImpact;
+    }
+
+    /// @notice Sets the slippage tolerance for a swap transaction.
+    /// @param _slippageTolerance Slippage tolerance in basis points. Value of
+    ///        this parameter must be between 0 and 10000 (inclusive). It
+    ///        should be chosen carefully as transactions with high slippage
+    ///        tolerance result in poor execution prices. On the other hand,
+    ///        very low slippage tolerance may cause transactions to be
+    ///        reverted frequently. Slippage tolerance equal to 0 is possible
+    ///        and disallows any slippage to happen on the swap at the cost
+    ///        of higher revert risk.
+    function setSlippageTolerance(uint256 _slippageTolerance)
+        external
+        onlyOwner
+    {
+        require(
+            _slippageTolerance <= BASIS_POINTS_DIVISOR,
+            "Maximum value is 10000 basis points"
+        );
+        slippageTolerance = _slippageTolerance;
+    }
+
+    /// @notice Sets the deadline for a swap transaction.
+    /// @param _swapDeadline Swap deadline in seconds. Value of this parameter
+    ///        should be grater than 0. It should be chosen carefully as
+    ///        transactions with long deadlines may result in poor execution
+    ///        prices. On the other hand, very short deadlines may cause
+    ///        transactions to be reverted frequently, especially in a
+    ///        gas-expensive environment. Deadline equal to 0 will effectively
+    //         make swaps impossible.
+    function setSwapDeadline(uint256 _swapDeadline) external onlyOwner {
+        swapDeadline = _swapDeadline;
+    }
+
+    /// @notice Swaps signer bonds on Uniswap v2 exchange.
+    /// @dev Swaps the given ETH amount for the target token using the Uniswap
+    ///      decentralized exchange. The maximum ETH amount is capped by the
+    ///      contract balance. Some governance parameters are applied on the
+    ///      transaction. The swap's price impact must fit within the
+    ///      maximum allowed price impact and the transaction is constrained
+    ///      with the slippage tolerance and deadline. Acquired target tokens
+    ///      are sent to the recipient address set during contract construction.
+    /// @param amount Amount to swap.
+    function swapSignerBondsOnUniswapV2(uint256 amount) external {
```

Comment:
Did not have a chance to dig in and it is not clear from the docs: is this function safe for "concurrent" use? Say there is 100 ETH to swap, then I execute a couple of transactions, one after another, and they all take some time to process on the Uniswap side. When they are all finally processed, successfully or not, does the state stay consistent?

Reply:
What do you mean by saying `does the state stay consistent?`? Regarding the rest, this is why we have a slippage tolerance parameter. Suppose three transactions: A, B, C. If they are sent one after another in that order, each of them will cause a specific price change (which depends on their price impact). If for example, A and B caused such a change that exceeds the slippage tolerance of C, C will be reverted.

Reply:
I mean the state of this contract - the amount of ETH it holds and so on. Say that there are 5 swaps pending, there is some outstanding ETH on this contract, three swaps succeed, two swaps fail and ETH comes back to this contract. Haven't had a chance to dig in so not saying it does not work, just asking :)

Reply:
Taking funds from this contract and executing a swap on Uniswap happens in the same transaction (Uniswap's method is payable). If something bad occurs during the swap, the whole transaction reverts, and funds remain untouched. 

Regarding multiple ones, if we have the following scenario:
1. Contract balance is `100 ETH`
2. Transaction A which swaps `80 ETH` is pending
3. Transaction B which swaps `90 ETH` is sent after transaction A but mined before them
4. `90 ETH` should be swapped and transaction A should revert due to insufficient balance. The final contract balance should be `10 ETH`

If we have:
1. Contract balance is `100 ETH`
2. Transaction A which swaps `40 ETH` is pending
3. Transaction B which swaps `50 ETH` is sent after transaction A but mined before them
4. Transaction A reverts due to, for example, slippage tolerance violation
5. `50 ETH` should be swapped and `40 ETH` from A should be untouched. The final contract balance should be `50 ETH`